packages/contract/src/Account.sol

```diff
@@ -163,4 +163,45 @@ contract Account is BaseAccount, UUPSUpgradeable, Initializable {
         (newImplementation);
         _onlySelf();
     }
+
+    /**
+     * Fetch all current signing keys
+     */
+    function getSigningKeys() public view returns (bytes32[2][] memory) {
+        return accountKeys;
+    }
+
+    /**
+     * Add a signing key to the account
+     * @param accountPubKey the P256 public key to add
+     */
+    function addSigningKey(bytes32[2] calldata accountPubKey) public onlySelf {
+        require(accountKeys.length < 255, "too many keys");
+        accountKeys.push(accountPubKey);
+        emit SigningKeyAdded(this, accountPubKey);
+    }
+
+    /**
+     * Remove a signing key from the account
+     * @param accountPubKey the P256 public key to remove
+     */
+    function removeSigningKey(
+        bytes32[2] calldata accountPubKey
+    ) public onlySelf {
+        require(accountKeys.length > 1, "cannot remove singular key");
+        for (uint256 i = 0; i < accountKeys.length; i++) {
+            if (
+                accountKeys[i][0] == accountPubKey[0] &&
+                accountKeys[i][1] == accountPubKey[1]
+            ) {
+                for (uint256 j = i;j < accountKeys.length - 1; j++) {
+                    accountKeys[j] = accountKeys[j + 1];
+                }
+                accountKeys.pop();
+                emit SigningKeyRemoved(this, accountPubKey);
+                return;
+            }
+        }
```

Comment:
looks great

one alternative to consider

- addSigningKey, removeSigningKey both take (index, pubKey)
- addSigningKey asserts `accountKeys[index]` is 0, then sets to pubKey
- removeSigningKey asserts `accountKeys[index]` equals pubKey, then sets to 0
- either way, assert index is in range 0-255
- track numSigningKeys, removeSigningKey asserts `numSigningKeys > 1`

advantages:
- both are now constant time/constant gas
- lets us refer to Device A/B/C or 1/2/3 without "Device C" changing into "Device B" when you remove B
- fixes an edge-case: if Device C has bad or no internet & doesn't yet know that Device B has been removed, and you sign an offline transaction from Device C (see Vitalik's "personal experiences" post for scenario), then that transaction will be still be valid (random nonce + stable keyIndex allows this to work)

Reply:
i recommend adding `(index, pubKey)` to the `SigningKeyAdded/Removed` event too

Reply:
done both